code/modules/nodes/helpers.dm

```diff
@@ -0,0 +1,109 @@
+
+//A file containing helpers for nodes and direction related things
+
+GLOBAL_LIST_EMPTY(allnodes)
+
+//If the input dir is diagonal, returns TRUE, otherwise FALSE
+/proc/IsDiagonal(direct)
+	if(direct in GLOB.cardinals)
+		return FALSE
+	return TRUE
+
+//Converts input direction to a list of either two same directions if the input is cardinal
+//Otherwise will return the two directions that make up the diagonal
+/proc/DiagonalToCardinal(direct)
+	if(direct in GLOB.cardinals)
+		return list(direct, direct)
+	return list(NSCOMPONENT(direct), EWCOMPONENT(direct))
+
+	switch(direct)
+		if(NORTHEAST)
+			return shuffle(list(NORTH, EAST))
+		if(NORTHWEST)
+			return shuffle(list(NORTH, WEST))
+		if(SOUTHEAST)
+			return shuffle(list(SOUTH, EAST))
+		if(SOUTHWEST)
+			return shuffle(list(SOUTH, WEST))
+
+//Returns a node that is in the direction of this node; must be in the src's adjacent node list
+/obj/effect/AINode/proc/GetNodeInDirInAdj(dir)
+
+	if(!datumnode.adjacent_nodes || !datumnode.adjacent_nodes.len)
+		return null
+
+	for(var/obj/effect/AINode/node in datumnode.adjacent_nodes)
+		if(get_dir(src, node) == dir)
+			return node
+	return null
+
+//Current weights are ENEMY_PRESENCE & DANGER_SCALE
+//Parameter call example
+//GetBestAdjNode(list(ENEMY_PRESENCE = -100, DANGER_SCALE = -1))
+//This means that the proc will pick out the *best* node, score is negatively impacted by amount of enemies and danger scale
+
+/obj/effect/AINode/proc/GetBestAdjNode(list/weight_modifiers)
+	if(weight_modifiers)
+		var/obj/effect/AINode/node_to_return = datumnode.adjacent_nodes[1]
+		var/current_best_node = 0
+		var/current_score = 0
+		for(var/obj/effect/AINode/node in shuffle(datumnode.adjacent_nodes)) //We keep a score for the nodes and see which one is best
+			for(var/i = 1; i != weight_modifiers.len; i++)
+				current_score += (node.datumnode.get_weight(i) * weight_modifiers[i])
+
+			if(current_score >= current_best_node)
+				current_best_node = current_score
+				node_to_return = node
+			current_score = 0
+
+		if(node_to_return)
+			return node_to_return
+
+	else //No weight modifier, return a adjacent random node
+		return pick(shuffle(datumnode.adjacent_nodes))
+
+//The equivalent of get_step_towards but now for nodes; will NOT intelligently pathfind based on node weights or anything else
+//Returns nothing if a suitable node in a direction isn't found, otherwise returns a node
+/proc/get_node_towards(obj/effect/AINode/startnode, obj/effect/AINode/destination)
```

Comment:
```suggestion
/datum/action_state/proc/get_node_towards(obj/effect/AINode/startnode, obj/effect/AINode/destination)
	if(startnode == destination)
		return startnode
	if(!length(startnode.datumnode.adjacent_nodes))
		return null
	
	var/node_dir = get_dir(startnode, destination)
	
	if(ISDIAGONALDIR(node_dir))
		var/secondary_dir
		if(prob(50))
			secondary_dir = node_dir & NORTH|SOUTH
			node_dir = node_dir & WEST|EAST
		else
			secondary_dir = node_dir & WEST|EAST
			node_dir = node_dir & NORTH|SOUTH
		for(var/n in startnode.datumnode.adjacent_nodes)
			var/obj/effect/AINode/node = n
			var/dir_to_check = get_dir(startnode, node)
			if(dir_to_check != node_dir && dir_to_check != secondary_dir)
				continue
			return node
	else
		for(var/n in startnode.datumnode.adjacent_nodes)
			var/obj/effect/AINode/node = n
			if(get_dir(startnode, node) != node_dir)
				continue
			return node
```
Depending on the behavior you want to achieve, you could simply do something like this instead:
```
for(var/n in startnode.datumnode.adjacent_nodes)
		var/obj/effect/AINode/node = n
		if(!(get_dir(startnode, node) & node_dir))
```
Then you wouldn't have to worry on if it's diagonal or not. The issueis that you wouldn't restrict it to cardinal adjacencies. So as said, depends on what you are after.

With this you could delete `DiagonalToCardinal` and `GetNodeInDirInAdj`